docs/proposals/20220927-label-sync-between-machine-and-nodes.md

```diff
@@ -0,0 +1,131 @@
+
+#### Story 1
+
+As a cluster admin/user, I would like a declarative and secure means by which to assign roles to my nodes. For example, when I do `kubectl get nodes`, I want the ROLES column to list my assigned roles for the nodes.
+
+#### Story 2
+
+As a cluster admin/user, for the purpose of workload placement, I would like a declarative and secure means by which I can add/remove labels on groups of nodes. For example, I want to run my outward facing nginx pods on a specific set of nodes (e.g. md-secure-zone) to reduce exposure to the company's publicly trusted server certificates.
+
+#### Story 3
+
+As a cluster admin/user, I want that Cluster API label management on Kubernetes Nodes do not conflits with labels directly managed by users or by other controllers.
+
+### Implementation Details/Notes/Constraints
+
+While designing a solution for syncing labels between Machine and underlying Kubernetes Nodes two main concerns have been considered:
+
+- Security, because Node labels can be used to attract workload to a specific Machine
+- Impacts on other components applying labels on Kubernetes Nodes, like e.g. Kubeadm, CPI etc.
+
+### Label Prefixes
+
+The idea of scoping syncronization to a well defined set of labels is a first answer to security/concurrency concerns; labels to be managed by Cluster API have been selected based on following criteria:
+
+- The `node-role.kubernetes.io` label has been used widely in the past to identify the role of aKubernetes Node (e.g. `node-role.kubernetes.io/worker=''`). For example, `kubectl get node` looks for this specific label when displaying the role to the user.
+
+- The `node-restriction.kubernetes.io/` prefix is recommended in the [Kubernetes docs](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#noderestriction) for things such as workload placement; according to https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/ also
+  labels like `example.node-restriction.kubernetes.io/fips=true` falls in this category (prefix =~ domain).
+
+- Cluster API will define its own label prefix, `node.cluster.x-k8s.io/`; for consistency we consider also `example.node.cluster.x-k8s.io/gpu=true` in this category (prefix =~ domain).
+
+#### Synchronization of CAPI Labels
+
+Only the labels that fall under the prefix described above will need to be kept in sync between the Machine and Node objects.
+
+The syncronization process is going to use [server side apply](https://kubernetes.io/docs/reference/using-api/server-side-apply/) in order to ensure that Cluster API will manage only the subset of labels coming from Machine objects and ignore labels applied concurrently by other users/other components.
+
+This requires to define an identity/manager name to be used by CAPI when performing this operation; additionally during implementation we are going to identify and address eventual additional steps required to properly transition existing Nodes to SSA, if required.
```

Comment:
Q: Should we mention something about SSA <=> workload cluster Kubernetes version considerations (e.g. only do it for Kubernetes >= v1.22 or alternatives)

Or is this part of "transition existing Nodes to SSA"?

Reply:
@enxebre opinions?

Reply:
[Context is ssa is enabled by default and beta since 1.16](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) and [better owner tracking was added in 1.18](https://kubernetes.io/blog/2020/04/01/kubernetes-1.18-feature-server-side-apply-beta-2/#wasn-t-it-already-beta-before-1-18) @sbueringer can you re-elaborate the concerns?

Reply:
We had issues when implementing Cluster topology reconciliation with SSA. IIRC the behavior changed a bit somewhere before v1.20, which is why we eventually restricted ClusterClass to mgmt clusters >=v1.22.

I'm not sure if those issue where with our final implementation or something we used before which was inspecting the managed fields (and their format changed between releases (?)).

@fabriziopandini Do you remember details?

It might be fine for >=v1.18 as we don't care in detail about how SSA uses managed fields


(cc @g-gaston that's the discussion around SSA I mentioned on the CRS proposal)

Reply:
For this effort, I think we can start by the assumption that SSA with all the supported releases will just work and do not introduce any new limitations on supported Kubernetes version upfront. We can eventually come back to this if tests prove us wrong. 

Note: The limitation on the management cluster version originated from the idea of relying on SSA internals / managed fields and results of tests surfacing some inconsistences in the older releases; then, when we finally moved to SSA dry-run, we dropped almost all those assumptions and now we mostly rely on the API server (there is only one place left where we patch managed fields).